### ISSUE TITLE
Implementing a two-layer neural network from scratch

### ISSUE THREAD

**Issue by ljvmiranda921** [ljvmiranda921/comments.ljvmiranda921.github.io#5]
Written on 02/16/2017 04:06:50 

URL: https://ljvmiranda921.github.io/notebook/2017/02/17/artificial-neural-networks/

**Comment by ljvmiranda921** [comment]
Comment written by **Vinod Kumar** on **07/20/2017 07:16:51**

<p>Thanks for the posts. Your posts have been very helpful.</p><p>Could you please help me with these confusions?</p><p>1. In these expressions:</p><p>grads['W2'] = <a href="http://np.dot" rel="nofollow noopener" title="np.dot">np.dot</a>(a1.T, dscores)</p><p>grads['W1'] = <a href="http://np.dot" rel="nofollow noopener" title="np.dot">np.dot</a>(X.T, dhidden)</p><p>Why is there a Transpose(a1.T and X.T)? Is it to make it suitable for dot product?</p><p>2. Why are we dividing the dscores by number of examples?<br>     dscores /= N</p>

**Comment by ljvmiranda921** [comment]
Comment written by **Lj Miranda** on **07/20/2017 16:28:35**

<p>Hi Vinod! Glad my posts helped!</p><p>As for your questions:</p><p>1. Yes you are correct. We want our matrices to have the same "inner size" so that doing a dot product won't break. Try checking the shape of the matrices in your implementation and make sure that $(a, b) \dot (b, c)$.</p><p>2. Recall that we are computing for a loss with the  following equation: <br>$$<br>L = (1/N) \sum_{i} L_{i} + reg_term<br>$$</p><p>The derivation of the softmax doesn't include the 1/N yet, that's why we are<br>dividing it after computation.</p>

**Comment by ljvmiranda921** [comment]
Comment written by **Vinod Kumar** on **07/20/2017 20:55:02**

<p>Oh right! I was only looking at the derivative of the loss function of a single example (Li) instead of the total loss.</p><p>Thanks for the clarification :)</p>

**Comment by ljvmiranda921** [comment]
Comment written by **chris** on **07/22/2017 13:58:43**

<p>I might just jump in here since my question is basically one step before Vinod's questions:</p><p>Why are we doing this:</p><p>dscores[range(N),y] -= 1</p><p>instead of:</p><p>dscores[range(N),y] = -1</p><p>Why are we subtracting one for the right class score instead of setting it to '-1'.<br>From the lecture slides of cs231n I assumed we were basically doing df/df as the very beginning of backprob, so in this particular case this would be -(dL_i/dL_i). wouldn't it?</p>

**Comment by ljvmiranda921** [comment]
Comment written by **Lj Miranda** on **07/22/2017 14:54:43**

<p>Hi Chris,</p><p>When we're doing backprop, one of the first steps would be to compute the gradient of the loss L_i with respect to the output f_k. Intuitively, we're "measuring" the change in the loss with respect to the output of the network, that is, dL_i / df_k . It turns out that the answer for this is simply</p><p>(dL_i / df_k) = p_k - 1 , where p_k is the softmax output (or the scores in our code).</p><p>Anyways, what we're gonna do now is to compute (dL_i / df_k) .</p><p>This can actually be seen in the lecture notes here (<a href="http://cs231n.github.io/neural-networks-case-study/#grad)" rel="nofollow noopener" title="http://cs231n.github.io/neural-networks-case-study/#grad)">http://cs231n.github.io/neu...</a>.</p><p>For the actual derivation, what they did is that they set the softmax output as p (as p_k = exp(f_k) / sum(exp(f_j)) ) so that L_i = -log(p_k). And then via chain rule:</p><p>dL_i / df_k = (dL_i / dp_k) * (dp_k / df_k)</p><p>The first term is just a simple derivative of the log() function. The second term is the derivative of the softmax function. The math in the second term is a bit involved (and I don't know how to typeset LaTeX here in Disqus comments. It may look messy) so I can refer you to this link:  <a href="http://eli.thegreenplace.net/2016/the-softmax-function-and-its-derivative/" rel="nofollow noopener" title="http://eli.thegreenplace.net/2016/the-softmax-function-and-its-derivative/">http://eli.thegreenplace.ne...</a></p><p>Hope this explanation helped!</p>

**Comment by ljvmiranda921** [comment]
Comment written by **chris** on **07/23/2017 21:25:51**

<p>Hi!</p><p>Yes, thanks a lot for the detailed explanation. I'm trying to map the calculations to the circuits that are visualized on the slides. However, I'm still struggling which gates are needed and which can be discarded and packed together with others.</p><p>Sorry for bothering, but could you briefly explain this step?:</p><p>dhidden = <a href="http://np.dot" rel="nofollow noopener" title="np.dot">np.dot</a>(dscores, W2.T)</p><p>Why are we using W2? What is the formula behind it?</p><p>Thanks a lot again!!</p>

**Comment by ljvmiranda921** [comment]
Comment written by **Matan Levy** on **08/25/2017 10:29:33**

<p>Lj Miranda, thanks for the detailed explanations!</p><p>I have two small questions though, <br>1) when computing the regularization loss should we also consider the bias vectors?  <br>2) for some reason I don't get the same loss as in the instructions (both with my code and when testing your), do you any idea what I am doing wrong?</p>

**Comment by ljvmiranda921** [comment]
Comment written by **Lj Miranda** on **08/25/2017 11:07:26**

<p>Hi Matan!</p><p>1. For regularization, remember that we are using L2-regularization, you can find the equation in the CS231n course notes (<a href="http://cs231n.github.io/neural-networks-2/#reg)" rel="nofollow noopener" title="http://cs231n.github.io/neural-networks-2/#reg)">http://cs231n.github.io/neu...</a>.</p><p>The intuition behind regularization is that we want our weights/parameters to be as "sparse" or as "simple" as possible. This means that we want to reduce complexity in our model and prevent it to overfit. So if we have a dense weight matrix, then we get a higher regularization loss.</p><p>Suppose we have a model that looks like this:</p><p>y = X0 + (W1 * X1) + (W2 * X2 ^ 2) + (W3 * X3 ^ 3) + (W4 * X4 ^ 4)</p><p>If we have a dense weight vector, that is, W = [1.0,  0.8, 0.74, 1.0], then most of the Xs are "turned on,"  the problem here is that there is a large tendency for this kind of model to overfit, and we don't want that. Now, if we optimize with a regularization loss to constrain our weights to be close to 0, then we might have a weight vector that looks like</p><p>W = [0.5, 0.001, 0.8, 0.00002]</p><p>The effect of W2 and W4 is lessened, and hopefully we arrive at a simpler model. The good thing about simpler models is that they tend to generalize better than overfit ones. And if they can generalize better, they can score higher if you feed it with data that is not seen.</p><p>Thus we see that regularization aims to increase bias in our model by penalizing our weights. Thus, the bias vectors have nothing to do with it, and so we don't really include in our computations.</p><p>2. As for #2, what are you getting? Hmmm, interesting. I will check on that again.</p>

**Comment by ljvmiranda921** [comment]
Comment written by **Lj Miranda** on **08/25/2017 11:09:08**

<p>Correction, in fact what's happening when we do regularization is that we try to avoid peaks in our weight vectors and try to achieve a more diffused weight vector.</p>

**Comment by ljvmiranda921** [comment]
Comment written by **Matan Levy** on **08/25/2017 14:11:19**

<p>Thank you!</p><p>1) I know better understand why we use the regularization loss,<br>I thought we should also include the bias, since in both SVM and Softmax classifiers we did apply the regularization loss on the bias as well (in these questions the bias came in the form of extra feature to the input).</p><p>2) I am not sure why, but I just had to omit the 0.5 factor from the regularization loss</p>

**Comment by ljvmiranda921** [comment]
Comment written by **earnshae** on **12/12/2018 03:11:17**

<p>In order to get the same answers you have above I have to uses the following regularization expressions for W2 and W1:</p><p>grads['W2'] += 2 * reg * W2<br>grads['W1'] += 2 * reg * W1</p><p>I think the code code you have posted above may discount the p=0.5 per layer regularization factor which must be divided out for the correct answer.</p><p>This correlates with this students work: <a href="https://github.com/rahul1990gupta/scs231n/blob/master/assignment1/two_layer_net.ipynb" rel="nofollow noopener" title="https://github.com/rahul1990gupta/scs231n/blob/master/assignment1/two_layer_net.ipynb">https://github.com/rahul199...</a></p>
